text/0000-ember-data-identifiers.md

```diff
@@ -0,0 +1,194 @@
+- Start Date: 2018-11-25
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Ember Data Record Identifiers 
+
+## Summary
+
+A follow-up RFC to [RecordData]() that simplifies APIs and ensures capabilities that were intended by the original RFC but not delivered.
+
+Specifically it ensures that `RecordData` can be implemented as a `singleton`, eliminates several redundant APIs that creeped into the original implementation, and simplifies the method signatures of `RecordData` APIs.
+
+The primary shift is that we would now utilize `json-api` resource identifiers and upgrade them within the store to `RecordIdentifier`s to communicate `type` `id` and `lid` information between APIs. We also deprecate use of `clientId` in favor of `lid`.
+
+## Motivation
+
+- nested saves / transactions
+- websockets
+- reload
+- better cache serialization
+- simpler cache and relationship management
+
+## Detailed design
+
+### New Typings
+
+```typescript
+import { Value as JSONValue } from 'json-typescript';
+
+export type Meta = { [k: string]: JSONValue };
+
+export interface ResourceIdentifier {
+  lid?: string;
+  id: string | null;
+  type: string;
+  meta?: Meta;
+}
+
+export interface RecordIdentifier {
+    lid: string;
+    id: string | null;
+    type: string;
+    meta: Meta | null;
+}
+```
+
+### New APIs.
+
+* **recordIdentifierFor**
+
+`import { recordIdentifierFor } from '@ember-data/store';`
+
+```typescript
+/*
+  Returns the RecordIdentifier for the given ResourceIdentifier.
+
+  Specifically this means that we:
+
+  - validate the `id` `type` and `lid` combo against known identifiers
+  - return an object that is guaranteed to have the `lid` for the given `type` and `id`
+  - return an object that also contains the stored `meta`
+  - this object is referentially stable
+*/
+export function recordIdentifierFor(store: Store, resourceIdentifier: ResourceIdentifier): RecordIdentifier {}
+```
+
+### Changes to existing APIs.
+
+**Store Methods**
+* `recordDataFor(modelName, id, clientId)`
+  * New Re-Arg: `recordDataFor(identifier)`
+* `createRecordDataFor(modelName, id, clientId, storeWrapper)`
+  * New Re-Arg: `createRecordDataFor(identifier, storeWrapper)`
+
+**StoreWrapper Methods**
+  * `recordDataFor(modelName, id, clientId)`
+    * New Re-Arg: `recordDataFor(identifier)`
+  * `notifyPropertyChange(modelName, id, clientId, key)`
+    * New Re-Arg: `notifyPropertyChange(identifier, propertyName)`
+  * `notifyHasManyChange(modelName, id, clientId, key)`
+    * deprecated (not in original RFC) in favor of `notifyPropertyChange`
+  * `notifyBelongsToChange(modelName, id, clientId, key)`
+    * deprecated (not in original RFC) in favor of `notifyPropertyChange`
+  * `attributesDefinitionFor(modelName)`
+    * New Re-Arg: `attributesDefinitionFor(identifier)`
+  * `relationshipsDefinitionFor(modelName)`
+    * New Re-Arg: `relationshipsDefinitionFor(identifier)`
+  * `setRecordId(modelName, id, clientId)`
```

Comment:
Should we make this just `notifyPropertyChange` `id`, no reason to make set id special?

Reply:
It is special as we use this to notify that the ID has been received from a save when no ID was present before and that we must now update the cache key / index.

The data flow is *whacked* here though, as `store._setRecordId` which this delegates to is going to call `internalModel.setId` which needs to be fixed to call `recordData.<updateId>` fpr the case where someone calls `store.createRecord('foo', {}).set('id', '<someid>');`